storage: avoid deadlock on (*Stopper).Quiesce

tbg opened the pull request:
See
https://github.com/cockroachdb/cockroach/issues/37954#issuecomment-498451871.

Closes #37954.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37986)
<!-- Reviewable:end -->

nvanbenschoten commented:
Nice find! I would have expected the fix to be selecting on `p.cfg.Stopper.ShouldQuiesce()` in `Provider.Subscribe` though, [here](https://github.com/cockroachdb/cockroach/blob/5a17b3dcee1f6d7db9b87e705798bac80594b27c/pkg/storage/closedts/provider/provider.go#L276) and [here](https://github.com/cockroachdb/cockroach/blob/5a17b3dcee1f6d7db9b87e705798bac80594b27c/pkg/storage/closedts/provider/provider.go#L325).

tbg committed:
storage: avoid deadlock on (*Stopper).Quiesce

See
https://github.com/cockroachdb/cockroach/issues/37954#issuecomment-498451871.

Closes #37954.

Release note: None

tbg commented:
Good point, I somehow thought we didn't have a stopper. Updated.

tbg commented:
bors r=nvanbenschoten
